contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,643 @@
+    uint constant RAY = 10 ** 27;
+
+    // This Struct tracks position information for a specific investor address
+    struct Investor {
+        uint256 layers;          // Number of CDP layers. According to github schematic
+        uint256 principalETH;    // WAD - principalETH contribution in Eth
+        uint256 collatRatio;     // WAD - Collaterazation ratio
+        bytes32 cdpID;           // cdpID
+        uint256 daiAmountFinal;  // WAD - Amount of Dai that an investor has after leveraging
+        uint256 drawnDai;        // WAD - Total amount of DAI drawn
+        uint256 priceFloor;      // WAD - Price floor in ETH
+        uint256 index;           // index of investor
+        uint256 purchPrice;      // WAD; currPrice at time of purchase
+    }
+    mapping (address => Investor) public investors;
+    address[] public investorAddresses;
+
+
+    // Modifiers
+    modifier onlyInvestor {
+        require(investorAddresses.length != 0, "Investor Array is empty");
+        require(investorAddresses[investors[msg.sender].index] == msg.sender, "Investor has not opened a position yet");
+        _;
+    }
+
+
+    // Events
+    event OpenPosition(address owner, uint256 principalETH, uint256 purchPrice, uint256 layers);
+    event ClosePosition(address owner, uint256 sellPrice);
+
+
+    /**
+    * @dev Constructor function
+    * Collects module information from the tub and matchingMarket (OasisDex)contracts
+    */
+    constructor()
+        public
+    {
+        address _tub = 0xa71937147b55Deb8a530C7229C442Fd3F31b7db2;         //Kovan
+        address _oasisDex = 0x8cf1Cab422A0b6b554077A361f8419cDf122a9F9;    //Kovan
+        tub = ITub(_tub);
+        weth = tub.gem();
+        peth = tub.skr();
+        dai = tub.sai();
+        pip = tub.pip();
+        pep = tub.pep();
+        gov = tub.gov();
+        fee = tub.fee();
+        axe = tub.axe();
+        dex = IMatchingMarket(_oasisDex);
+
+        makerLR = ray2wad(tub.mat());
+
+        // we approve tub, tap, and dex to access weth, peth, and dai contracts
+        weth.approve(tub, uint(-1));
+        weth.approve(dex, uint(-1));
+        peth.approve(tub, uint(-1));
+        dai.approve(tub, uint(-1));
+        dai.approve(dex, uint(-1));
+        gov.approve(tub, uint(-1));
+        gov.approve(dex, uint(-1));
+    }
+
+
+    // Fallback function
+    function() payable public {}
+
+
+    /**
+    * @dev Leverage - Creates a leveraged long USD/ETH position
+    * principalETH (message value) of sender is used to:
+    * 1. Open CDP
+    * 2. Lock PETH into a CDP
+    * 3. Draw DAI and trade for WETH on OasisDEX
+    * 4. Convert WETH to PETH
+    * 5. Repeat steps 2 - 4 for # of (layers-1) prescribed
+    * 6. Holds CDP ownership and remaining DAI
+    * @param _priceFloorETH - position liquidated if USD/ETH price drops below this point
+    * @param _layers - # of layers used to leverage
+    */
+    function leverage(uint256 _priceFloorETH, uint256 _layers)
```

Comment:
_layers is still unbounded. We'll end up running out of gas in the trade loop if somebody calls it with too many layers.